command-expansion-server/src/getActivityTypescript.ts

```diff
@@ -0,0 +1,131 @@
+import {GraphQLClient, gql} from 'graphql-request';
+import {globalDeclaration, indent, interfaceDeclaration} from "./packages/lib/CodegenHelpers.js";
+import {ErrorWithStatusCode} from "./utils/ErrorWithStatusCode.js";
+
+enum SchemaTypes {
+  Int = 'int',
+  Real = 'real',
+  Duration = 'duration',
+  Boolean = 'boolean',
+  String = 'string',
+  Series = 'series',
+  Struct = 'struct',
+  Variant = 'variant',
+}
+
+type Schema =
+  | IntSchema
+  | RealSchema
+  | DurationSchema
+  | BooleanSchema
+  | StringSchema
+  | SeriesSchema
+  | StructSchema
+  | VariantSchema
+  ;
+
+interface BaseSchema<T extends SchemaTypes | unknown> {
+  type: T,
+}
+
+type StringSchema = BaseSchema<SchemaTypes.String>;
+type BooleanSchema = BaseSchema<SchemaTypes.Boolean>;
+type IntSchema = BaseSchema<SchemaTypes.Int>;
+type RealSchema = BaseSchema<SchemaTypes.Real>;
+type DurationSchema = BaseSchema<SchemaTypes.Duration>;
+
+interface SeriesSchema extends BaseSchema<SchemaTypes.Series> {
+  items: Schema,
+}
+
+interface StructSchema extends BaseSchema<SchemaTypes.Struct> {
+  items: {
+    [key: string]: Schema,
+  }
+}
+
+interface VariantSchema extends BaseSchema<SchemaTypes.Variant> {
+  variants: {
+    key: string,
+    label: string,
+  }[]
+}
+
+interface GraphQLActivityParameter {
+  order: number,
+  schema: Schema,
+}
+
+interface GraphQLActivity {
+  name: string,
+  parameters: GraphQLActivityParameter[],
+  requiredParameters: string[],
+}
+
+export async function getActivityTypescript(
+  graphqlClient: GraphQLClient,
+  missionModelId: number,
+  activityTypeName: string
+): Promise<string> {
+  console.log(`query parameters from ${activityTypeName} with missionModelId: ${missionModelId}`);
+  const response = await graphqlClient.request<{
+    activity_type: GraphQLActivity[]
+  }>(gql`
```

Comment:
Maybe this is a little pedantic but the mixing of GraphQL querying and code-gen. logic within the same "translation unit" doesn't feel quite right. I can understand how decoupling this querying from generation may feel be a bit overkill right now but in my experience having disciplined decoupling early on can save trouble later.

Reply:
Another benefit to decoupling would be testability. Currently, in order to test this code gen. you'd need to have `GraphQLClient`. Instead, if there was an abstract interface that the code generation could take as input then you could write a series of tests fairly easily.

Reply:
That's fair enough, all boils down to personal preference about modularity.

>and incurs runtime costs

I'd imagine a call out to another class/function should be very cheap (<10 ms.?). Either way, all good :)